### ISSUE TITLE
Add command-completion for PowerShell invocation command line

### ISSUE THREAD

**Issue by Liturgist** [PowerShell/PowerShell#14808]
## Summary of the new feature/enhancement
Command-completion (tab-completion) is a nice capability It would be helpful to have command-completion for parameters when invoking PowerShell from a PowerShell console.

**Comment by theJasonHelmick** [comment]
Hi @Liturgist, I want to verify I understand your feature request, from within a PowerShell console or terminal running PowerShell, you want tab-completion to complete the pwsh.exe syntax below?
pwsh[.exe] [-Login] [[-File] <filePath> [args]]
                  [-Command { - | <script-block> [-args <arg-array>]
                                | <string> [<CommandParameters>] } ]
                  [-ConfigurationName <string>] [-CustomPipeName <string>]
                  [-EncodedCommand <Base64EncodedCommand>]
                  [-ExecutionPolicy <ExecutionPolicy>] [-InputFormat {Text | XML}]
                  [-Interactive] [-MTA] [-NoExit] [-NoLogo] [-NonInteractive] [-NoProfile]
                  [-OutputFormat {Text | XML}] [-SettingsFile <filePath>] [-SSHServerMode] [-STA]
                  [-Version] [-WindowStyle <style>] [-WorkingDirectory <directoryPath>]

**Comment by SeeminglyScience** [comment]
@theJasonHelmick I'm pretty certain you're right and that's what they're asking for.  A build in argument completer for it would be great, I always found it strange that there wasn't one.

**Comment by Liturgist** [comment]
@theJasonHelmick, yes, I believe you understand correctly. My apologies for inattention. I still think this would be a helpful addition.

**Comment by theJasonHelmick** [comment]
@Liturgist - Thank you for the response and the suggestion -- I think this could be a benefit to customers unfamiliar with pwsh.exe syntax.  I'll discuss further with the working group and report back.

**Comment by Liturgist** [comment]
I checked code in a few times, but always had to resolve some issue. I could never overcome a rebase, so gave up.

**Comment by iSazonov** [comment]
We could add completer for pwsh.exe

**Comment by theJasonHelmick** [comment]
@Liturgist @iSazonov -- After discussing this with the working group, we agree that this would be a beneficial feature for users.  I'm marking this up-for-grabs if the community would like to consider building this solution.

**Comment by powercode** [comment]
Would it be OK to add a reference to System.CommandLine?

**Comment by SeeminglyScience** [comment]
> Would it be OK to add a reference to System.CommandLine?

Any reference we add essentially version locks that library for anyone else who wants to consume it. There's a pretty high bar for impact that needs to be hit for that to be considered

**Comment by Liturgist** [comment]
This is still the case. Just waiting for someone to take it on.

PS C:\> $psversionTable.psversion.ToString()
7.3.9

**Comment by Liturgist** [comment]
This is still the case.

PS C:\> $PSVersionTable.PSVersion.ToString()
7.4.2

**Comment by jshigetomi** [comment]
@Liturgist I'll take this on. Let me know if someone is already working on it.

**Comment by jshigetomi** [comment]
I made an argument completer for pwsh!

https://gist.github.com/jshigetomi/58cb79de086c804cda3173ef63d7dddf

Please leave some comments and ideas. I hope we can incorporate this completer during runtime once it's approved.

**Comment by Liturgist** [comment]
@jshigetomi, I will look forward to trying it out. Thanks.

**Comment by Liturgist** [comment]
@jshigetomi, many thanks for your work. 

After running PwshTabCompletion.ps1, the completer produces the following parameter names.

`-File, -f, -Command, -c, -CommandWithArgs, -cwa, -ConfigurationName, -config, ...`

I was expecting...

`-File, -Command, -ConfigurationFile, -ConfigurationName, CustomPipeName, ...`

Where does CommandWithArgs come from? It does not appear to be in the pwsh help.

```
    Usage: pwsh[.exe]
        [-Login]
        [[-File] <filePath> [args]]
        [-Command { - | <script-block> [-args <arg-array>]
                      | <string> [<CommandParameters>] } ]
        [-ConfigurationFile <filePath>]
        [-ConfigurationName <string>]
        [-CustomPipeName <string>]
        [-EncodedCommand <Base64EncodedCommand>]
        [-ExecutionPolicy <ExecutionPolicy>]
        [-InputFormat {Text | XML}]
        [-Interactive]
        [-MTA]
        [-NoExit]
        [-NoLogo]
        [-NonInteractive]
        [-NoProfile]
        [-NoProfileLoadTime]
        [-OutputFormat {Text | XML}]
        [-SettingsFile <filePath>]
        [-SSHServerMode]
        [-STA]
        [-Version]
        [-WindowStyle <style>]
        [-WorkingDirectory <directoryPath>]
```

**Comment by jshigetomi** [comment]
@Liturgist It's been awhile since this discussion but I think there's been meaningful headway on this topic
@daxian-dbw created this module that is compatible with 7.6.X.
https://github.com/daxian-dbw/PSNativeToolCompletion

Importing this module in combination with the below gist I rewrote will give you a nice solution for pwsh tab completion.

https://gist.github.com/jshigetomi/88395ec026f4c246cfa6f9df9f35e185

<blockquote><img src="https://opengraph.githubassets.com/8e4302be471c087a045caa691b78e1b6daa115b2404abcdeed60e2c58fde7dd4/daxian-dbw/PSNativeToolCompletion" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/daxian-dbw/PSNativeToolCompletion">GitHub - daxian-dbw/PSNativeToolCompletion: A PowerShell module that enables lazy loading of completion scripts for native commands.</a></strong></div><div>A PowerShell module that enables lazy loading of completion scripts for native commands. - daxian-dbw/PSNativeToolCompletion</div></blockquote>
<blockquote><img src="https://github.githubassets.com/assets/gist-og-image-54fd7dc0713e.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> Gist</div><div><strong><a href="https://gist.github.com/jshigetomi/88395ec026f4c246cfa6f9df9f35e185">pwsh tab completer with JSON metadata tooltips (for PSNativeToolCompletion module)</a></strong></div><div>pwsh tab completer with JSON metadata tooltips (for PSNativeToolCompletion module) - __pwsh.ps1</div></blockquote>

**Comment by jshigetomi** [comment]
I'm reopening this issue because pwsh still doesn't natively ship the completer. The above work around works reasonably well for now.
